File: generated-code/codecraft/csharp/Model/RepairProperties.cs
namespace TransGenTest.Model
{
    public struct RepairProperties
    {
        public Model.EntityType[] ValidTargets { get; set; }
        public int Power { get; set; }
    
        public RepairProperties(Model.EntityType[] validTargets, int power)
        {
            this.ValidTargets = validTargets;
            this.Power = power;
        }
    
        public static RepairProperties ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new RepairProperties();
            result.ValidTargets = new Model.EntityType[reader.ReadInt32()];
            for (int validTargetsIndex = 0; validTargetsIndex < result.ValidTargets.Length; validTargetsIndex++)
            {
                result.ValidTargets[validTargetsIndex] = EntityTypeHelper.ReadFrom(reader);
            }
            result.Power = reader.ReadInt32();
            return result;
        }
    
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(ValidTargets.Length);
            foreach (var validTargetsElement in ValidTargets)
            {
                writer.Write((int) (validTargetsElement));
            }
            writer.Write(Power);
        }
    
        public override string ToString() {
            string stringResult = "RepairProperties { ";
            stringResult += "ValidTargets: ";
            stringResult += "[ ";
            int validTargetsIndex = 0;
            foreach (var validTargetsElement in ValidTargets)
            {
                if (validTargetsIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += validTargetsElement.ToString();
                validTargetsIndex++;
            }
            stringResult += " ]";
            stringResult += ", ";
            stringResult += "Power: ";
            stringResult += Power.ToString();
            stringResult += " }";
            return stringResult;
        }
    }
}

File: generated-code/codecraft/fsharp/Model/RepairProperties.fs
#nowarn "0058"

namespace TransGenTest.Model

/// Entity's repair properties
type RepairProperties = {
    /// Valid target entity types
    ValidTargets: EntityType[];
    /// Health restored in one tick
    Power: int;
} with

    /// Write RepairProperties to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.ValidTargets.Length
        this.ValidTargets |> Array.iter (fun value ->
            writer.Write (int value) )
        writer.Write this.Power

    /// Read RepairProperties from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        ValidTargets = [|for _ in 1 .. reader.ReadInt32() do
                           yield reader.ReadInt32() |> enum |]
        Power = reader.ReadInt32()
    }